File: explicit-python/explicit_nlu/parser/xml/ExplicitXmlParser.py
from typing import Dict, List

from xml.dom import minidom
from xml.dom.minidom import Element, Text

from explicit_nlu.parser import ExplicitRule
from explicit_nlu.parser.AntlrParser import AntlrParser
from explicit_nlu.parser.ExplicitRules import ExplicitRules, Token


class ExplicitXmlParser:
    def parse(self, file_path: str) -> ExplicitRules:
        xdom = minidom.parse(file_path)
        root = xdom.childNodes[0]

        return ExplicitRules(root.tagName,
                             self._parse_rules(root),
                             self._parse_labels(root),
                             self._parse_tokens(root),
                             self._parse_mappings(root),
                             self._parse_patterns(root),
                             self._parse_features(root))

    def _parse_rules(self, root: Element) -> List[ExplicitRule]:
        result: List[ExplicitRule] = []
        rules = root.getElementsByTagName("rule")

        for rule in rules:
            eql = self._to_text(rule.getElementsByTagName("eql")[0])
            idx = [self._to_text(x) for x in rule.getElementsByTagName("idx")]
            ner: List[Dict[str, str]] = []

            for element in rule.getElementsByTagName("ner"):

                # explicit entities
                for entity in element.childNodes:
                    if isinstance(entity, Text):
                        continue
                    if entity.tagName != "entity":
                        continue

                    entity_map: Dict[str, str] = {}

                    if entity.hasAttribute("repeat"):
                        entity_map[":repeat"] = entity.getAttribute("repeat")
                    if entity.hasAttribute("foreach"):
                        entity_map[":foreach"] = entity.getAttribute("foreach")

                    for attribute in entity.childNodes:
                        entity_map[attribute.tagName] = self._to_text(attribute)

                    ner.append(entity_map)

                # implicit entities
                entity_map: Dict[str, str] = {}
                for entity in element.childNodes:
                    if isinstance(entity, Text):
                        continue
                    if entity.tagName == "entity":
                        continue

                    for attribute in entity.childNodes:
                        entity_map[entity.tagName] = self._to_text(attribute)

                if len(entity_map) > 0:
                    ner.append(entity_map)

            result.append(ExplicitRule(AntlrParser().get_tokens(eql), ner, list(map(lambda x: int(x), idx))))

        return result

    def _parse_labels(self, xdom: Element) -> Dict[str, str]:
        result: Dict[str, str] = dict()
        labels = xdom.getElementsByTagName("labels")

        for label in labels:
            for node in label.childNodes:
                if not isinstance(node, Text):
                    result[self._to_text(node)] = node.tagName

        return result

    def _parse_mappings(self, xdom: Element) -> Dict[str, str]:
        result: Dict[str, str] = dict()
        mappings = xdom.getElementsByTagName("mappings")

        for mapping in mappings:
            for node in mapping.childNodes:
                if not isinstance(node, Text):
                    result[node.getAttribute("inbound")] = node.getAttribute("outbound")

        return result

    def _parse_patterns(self, xdom: Element) -> Dict[str, str]:
        result: Dict[str, str] = dict()
        patterns = xdom.getElementsByTagName("patterns")

        for pattern in patterns:
            for node in pattern.childNodes:
                if not isinstance(node, Text):
                    result[node.tagName] = self._to_text(node)

        return result

    def _parse_features(self, xdom: Element) -> Dict[str, str]:
        result: Dict[str, str] = dict()
        patterns = xdom.getElementsByTagName("features")

        for pattern in patterns:
            for node in pattern.childNodes:
                if not isinstance(node, Text):
                    result[node.tagName] = self._to_text(node)

        return result

    def _parse_tokens(self, xdom: Element) -> List[Token]:
        result: List[Token] = list([])
        tokens = xdom.getElementsByTagName("tokens")

        for token in tokens:
            for node in token.childNodes:
                if not isinstance(node, Text):
                    if node.nodeName != "#text" and node.nodeName != "#comment":
                        result.append(Token(node.getAttribute("pattern"),
                                            node.getAttribute("replacement"),
                                            str(node.getAttribute("boundary")) != "false",
                                            str(node.getAttribute("regex")) == "true"))

        return result

    def _to_text(self, node: Element) -> str:
        if node.nodeType == node.TEXT_NODE:
            return node.data
        else:
            text_string = ""
            for child_node in node.childNodes:
                text_string += self._to_text(child_node)
            return text_string

File: explicit-java/src/main/kotlin/explicit/parser/xml/ExplicitXmlParser.kt
package explicit.parser.xml

import explicit.api.extension.*
import explicit.parser.AntlrParser
import explicit.parser.ExplicitRule
import explicit.parser.ExplicitRules
import org.w3c.dom.Node
import org.w3c.dom.NodeList
import java.io.InputStream
import java.util.regex.Pattern
import javax.xml.parsers.DocumentBuilderFactory

class ExplicitXmlParser {

    companion object {
        val DOCUMENT_BUILDER = ThreadLocal.withInitial {
            DocumentBuilderFactory.newInstance().newDocumentBuilder()
        }!!
    }

    private fun toMapOfPatterns(map: Map<*, *>): Map<String, Pattern> {
        val patterns = java.util.HashMap<String, Pattern>()
        map.forEach { k, v -> patterns[k.toString()] = Pattern.compile(v.toString()) }
        return patterns
    }

    fun parse(source: String) = parse(java.io.FileInputStream(source))

    fun parse(stream: InputStream): ExplicitRules {
        val document = DOCUMENT_BUILDER.get().parse(stream)
        val root = document.firstChild

        val labelMap = parseLabels(root.childNodes)
        val mappings = parseMappings(root.childNodes)
        val patterns = parsePatterns(root.childNodes)
        val features = parseFeatures(root.childNodes)
        val acronyms = parseTokens(root.childNodes)
        return ExplicitRules(root.nodeName, parseRules(root.childNodes), labelMap, acronyms, mappings, toMapOfPatterns(patterns), features)
    }

    fun parseRule(node: Node): ExplicitRule {
        val eql = node.findText("eql")
        val ner = ArrayList<Map<String, String>>()

        // explicit entities
        ner.addAll(node.find("ner")
                .childNodes
                .filter { e -> e.nodeName == "entity" }
                .map { e ->
                    val entity = HashMap<String, String>()
                    e.findAttr("repeat").ifPresent { entity[":repeat"] = it }
                    e.findAttr("foreach").ifPresent { entity[":foreach"] = it }

                    e.childNodes
                            .filter { x -> x.nodeName == "#text" }
                            .filter { x -> x.nodeName == "#comment" }
                            .forEach { x -> entity[node.nodeName] = x.textContent }
                    entity
                }.fold(ArrayList()) { list, e ->
                    list.add(e)
                    list
                })

        val nerEntries = node.find("ner")
                .childNodes
                .filter { e -> e.nodeName != "entity" && e.nodeName != "#text" }

        // implicit entities
        if (nerEntries.isNotEmpty()) {
            ner.add(nerEntries
                    .map { e ->
                        val entity = HashMap<String, String>()
                        e.childNodes
                                .filter { x -> x.nodeName == "#text" }
                                .forEach { x -> entity[e.nodeName] = x.textContent }
                        entity
                    }.reduce { a, b ->
                        a.putAll(b)
                        a
                    })
        }

        val idx = getIdx(node)
        return ExplicitRule(AntlrParser().getTokens(eql), ner, idx)
    }

    private fun parseRules(list: NodeList): List<ExplicitRule> {
        return list.filter {
            !listOf("#text", "#comment", "mappings", "patterns", "postprocessing", "labels",
                    "features", "tokens").contains(it.nodeName)
        }.map(this::parseRule)
    }

    private fun parseLabels(list: NodeList): Map<String, String> {
        return list.filter { it.nodeName == "labels" }
                .flatMap { it.childNodes.toList() }
                .filter { it.nodeName != "#text" }
                .filter { it.nodeName != "#comment" }
                .map { mapOf(it.textContent to it.nodeName) }
                .fold(HashMap()) { a, b ->
                    a.putAll(b)
                    a
                }
    }

    private fun parseMappings(list: NodeList): Map<String, String> {
        return list.filter { it.nodeName == "mappings" }
                .flatMap { it.childNodes.toList() }
                .filter { it.nodeName != "#text" }
                .map { mapOf(it.findAttr("inbound").get() to it.findAttr("outbound").get()) }
                .fold(HashMap()) { a, b ->
                    a.putAll(b)
                    a
                }
    }

    private fun parsePatterns(list: NodeList): Map<String, String> {
        return list.filter { it.nodeName == "patterns" }
                .flatMap { it.childNodes.toList() }
                .filter { it.nodeName != "#text" }
                .filter { it.nodeName != "#comment" }
                .map { mapOf(it.nodeName to it.textContent) }
                .fold(HashMap()) { a, b ->
                    a.putAll(b)
                    a
                }
    }

    private fun parseFeatures(list: NodeList): Map<String, String> {
        return list.filter { it.nodeName == "features" }
                .flatMap { it.childNodes.toList() }
                .filter { it.nodeName != "#text" }
                .filter { it.nodeName != "#comment" }
                .map { mapOf(it.nodeName to it.textContent) }
                .fold(HashMap()) { a, b ->
                    a.putAll(b)
                    a
                }
    }

    private fun parseTokens(list: NodeList): ArrayList<ExplicitRules.Token> {
        return list.filter { it.nodeName == "tokens" }
                .flatMap { it.childNodes.toList() }
                .filter { it.nodeName != "#text" }
                .filter { it.nodeName != "#comment" }
                .map {
                    val pattern = it.findAttr("pattern").get()
                    val replacement = it.findAttr("replacement").orElse(null)
                    val boundary = it.findAttr("boundary").map { it.toBoolean() }.orElse(true)
                    val regex = it.findAttr("regex").map { it.toBoolean() }.orElse(false)
                    ExplicitRules.Token(pattern, replacement, boundary, regex)
                }
                .fold(ArrayList()) { a, b ->
                    a.add(b)
                    a
                }
    }

    private fun getIdx(node: Node): List<Int> {
        val nodeList = node.childNodes
        for (i in 0 until nodeList.length) {
            if (nodeList.item(i).nodeName === "idx") {
                val value = nodeList.item(i).textContent
                return value.split(",").map(String::trim).map(Integer::parseInt)
            }
        }
        return ArrayList()
    }

}
